makenian.py

```diff
@@ -35,29 +36,62 @@ def read_nine_letter_words(dictionary_file):
     all_words = [x.strip() for x in dictionary]
     return nine_letter_words(all_words)
 
-def select_word(words):
-    return random.choice(words)
+def select_word(words, previous_puzzles):
+    word = random.choice(words).upper()
+    sorted_letters = ''.join(sorted(word))
 
-def find_puzzle(words):
-    word = select_word(words)
+    attempts = 0
+    while sorted_letters in previous_puzzles:
+        print(f'Skipping seen nian {word}, and selecting another one')
+        word = random.choice(words)
+        sorted_letters = ''.join(sorted(word))
+        attempts += 1
+
+        if attempts > 10:
+            print(f'Made 10 attempts at selecting nian, but found only puzzles that have already been used; will use {word} anyway')
+            return word
+
+    return word
```

Comment:
We might be able to simplify this a bit by changing to a for loop with an early return if the puzzle is new. I'll take a stab next time I have a few minutes to hack.